Automattic/wp-calypso
# Migrate package-lock.json to yarn.lock

### Changes proposed in this Pull Request

* This is the PR that replaces `package-lock.json` with `yarn.lock`, so it has the potential to change the dependency tree.

### Testing instructions

There is not a lot to review (unless you want to read `yarn.lock` changes), so I'd suggest to rely on comparing the trees to check that nothing has changed.

It is recommended two have two checkouts, one with master and the other with this branch.

In master:
- Checkout master (I used commit 109dd84caf to do this check)
- Install deps with `rm -fr node_modules && npm ci`
- Generate an effective tree in list mode `node_modules/.bin/effective-module-tree -o list > ../npm-tree

In this branch:
- Checkout this branch
- Install deps with `rm -fr node_modules && yarn`
- Generate an effective tree in list mode `node_modules/.bin/effective-module-tree -o list > ../yarn-tree

Once you have both trees:
- Do a diff (`diff npm-tree yarn-tree > tree.diff`)
- Open it in an editor and validate that the changes are limited to the list of differences below

### Notes

It is important to mention that the dependency tree when moving to yarn **will be different**: npm can generate dependency trees that can't be represented in yarn with 100% fidelity. Specifically, in npm a range (eg: `lib@^1.0.0`) can be satisfied with more than one package (eg: some parts of the tree could get `lib@1.0.1` while other parts could get `lib@1.0.2`). This is not possible with yarn: once a range is resolved to whatever version, that version will be used anywhere on the tree where the range was used (eg: `lib@^1.0.0` will always resolve to `lib@1.0.2`)

After analysing both trees, I found many differences but all of them can be explained by the note above. Also, all of them are only present in devDependencies, so that kind of reduces the risk of unexpected changes in production.

Specifically, the differences between both trees are:
- `dom-serializer@0.1.1` vs `dom-serializer@0.2.2`. Reason:
	- In NPM, dependency dependency `domutils@1.5.1 -> dom-serializer@0` is resolved with two different versions: `0.1.1` and `0.2.2`.
	- In YARN, it is always resolved with `0.2.2`

- `acorn@7.1.0` vs `acorn@7.1.1`. Reason:
	- In NPM, package `espree@6.1.2` depends on `acorn@^7.1.0` and gets `acorn@7.1.0`. Package `jsdom@15.2.1` depends on `acorn@^7.1.0` and gets `acorn@7.1.1`
	- In YARN, both packages get `acorn@7.1.1`

- `domutils@1.7.0` vs `domutils@1.5.1`
	- In NPM, dependency `htmlparser2@3.10.1 -> domutils@^1.5.1` is resolved with two different versions: `1.7.0` and `1.5.1`
	- In YARN, it is always resolved with `1.7.0`

- `safe-buffer@5.2.0` vs `safe-buffer@5.1.2`
	- In NPM, depenency `minipass@2.9.0 -> safe-buffer@^5.1.2` is resolved with two different versions: `5.1.2` and `5.2.0`
	- In YARN, it is always resolved with `5.1.2`

- `find-cache-dir@3.2.0` vs `find-cache-dir@3.3.1`
	- In NPM, package `moment-timezone-data-webpack-plugin@1.1.0` depends on `find-cache-dir@^3.0.0` and gets `find-cache-dir@3.2.0`. Package `@storybook/core@5.3.14` depends on `find-cache-dir@^3.0.0` and gets `find-cache-dir@3.3.1`
	- In YARN, both packages get `find-cache-dir@3.3.1`

- `vfile-message@2.0.2`
    - In NPM, `vfile-message@2.0.2` is bring in as a transitive dependency for `@wordpress/server-side-render`, and hoisted to the root. Then `@automattic/media-library` assumes it is there when compiling itself.
    - In YARN, `vfile-message@1.1.1` is the version getting hoisted to the root, so `@automattic/media-library` fails to compile. The solution is to have a explicit dependency on `vfile-message@2.0.2` on the yarn tree.
